vortex-array/src/array2.rs

```diff
@@ -0,0 +1,150 @@
+#![allow(dead_code)]
+use std::marker::PhantomData;
+use std::sync::{Arc, RwLock};
+
+use vortex_buffer::Buffer;
+use vortex_dtype::DType;
+use vortex_error::{vortex_bail, VortexError, VortexResult};
+
+use crate::stats::StatsSet;
+use crate::{ArrayMetadata, Canonical, Context};
+
+// Type-erased array.
+pub struct ArrayData {
+    encoding: &'static EncodingVTable,
+    dtype: Arc<DType>,
+    len: usize,
+    inner: InnerArrayData,
+}
+enum InnerArrayData {
+    Owned(OwnedArrayData),
+    Viewed(ViewedArrayData),
+}
+struct OwnedArrayData {
+    metadata: Arc<dyn ArrayMetadata>,
+    buffer: Option<Buffer>,
+    children: Arc<[ArrayData]>,
+    stats_map: Arc<RwLock<StatsSet>>,
+}
+struct ViewedArrayData {
+    flatbuffer: Buffer,
+    flatbuffer_loc: usize,
+    buffers: Arc<[Buffer]>,
+    // TODO(ngates): move this onto ArrayData?
+    ctx: Arc<Context>,
+}
+
+// The set of generic array functionality.
+// Again, it's nice to split this into multiple traits to allow implementation to be split
+// across multiple files, and for us to auto-derive some of the implementations.
+pub trait ArrayImpl {
+    fn as_array_data(&self) -> &ArrayData;
+    fn into_array_data(self) -> ArrayData;
+    fn into_canonical(self) -> VortexResult<Canonical>;
+    fn is_valid(&self, index: usize) -> VortexResult<bool>;
+}
+
+// Array implementation is supported for the type-erased ArrayData by dispatching via the encoding
+// VTable.
+impl ArrayImpl for ArrayData {
+    fn as_array_data(&self) -> &ArrayData {
+        self
+    }
+
+    fn into_array_data(self) -> ArrayData {
+        self
+    }
+
+    fn into_canonical(self) -> VortexResult<Canonical> {
+        (self.encoding.into_canonical)(self)
+    }
+
+    fn is_valid(&self, index: usize) -> VortexResult<bool> {
+        (self.encoding.is_valid)(self, index)
+    }
+}
+
+pub struct EncodingVTable {
+    id: &'static str,
+    into_canonical: fn(data: ArrayData) -> VortexResult<Canonical>,
+    is_valid: fn(data: &ArrayData, idx: usize) -> VortexResult<bool>,
```

Comment:
why use the type-erased thing? this is the EncodingVTable, but every implementation needs to check the encoding of the thing it's given?

Reply:
The vtable _must_ be type erased. You can't `impl Array for ArrayData` otherwise